Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean

```diff
@@ -361,6 +368,87 @@ lemma mul_right_injective_of_inv (A : Matrix m n α) (B : Matrix n m α) (h : A
 
 end InjectiveMul
 
+section vecMul
+
+variable [Fintype m] [DecidableEq m] {K : Type*} [Field K]
+
+
+theorem vecMul_surjective_iff_exists_left_inverse {A : Matrix m n α} :
+    Function.Surjective A.vecMul ↔ ∃ B : Matrix n m α, B * A = 1 := by
+  refine ⟨fun h ↦ ?_, fun ⟨B, hBA⟩ y ↦ ⟨B.vecMul y, by simp [hBA]⟩⟩
+  choose rows hrows using (h <| Pi.single · 1)
+  refine ⟨Matrix.of rows, Matrix.ext fun i j => ?_⟩
+  rw [mul_apply_eq_vecMul, one_eq_pi_single, ←hrows]
+  rfl
```

Comment:
After my golf, this proof now works on semirings (dropping both `-` and commutativity); can you try to generalize some of the other results in the same way?